### DISCUSSION TITLE
Monitoring unfinished checkouts

### DISCUSSION THREAD

**Discussion by rrrnld** [saleor/saleor#13866] [General]
Did somebody build an app to monitor unfinished checkouts? We just had a case of a customer not being able to validate their own address in Korea, and we would like some way to see if a checkout did not end in an order that was placed. Maybe this is also possible in the dashboard, but it's not immediately obvious to me.

Thanks for any help!

**EDIT** I just found this, which is the example app linked to from the saleor docs. https://github.com/saleor/saleor-app-abandoned-checkouts. It looks like it doesn't go past the step of setting up & configuring the app, i.e. the actual monitoring feature is missing.

**EDIT 2** As far as I can tell, there's no easy way to filter checkouts by not-being-assigned-to-an-order, is that correct? I tried crafting a GraphQL query (see below), but I don't find a way to exclude some checkouts by id from a result set.

```graphql
query FetchAbandonedCheckouts($after: String) {
  orders(first: 20 sortBy: { field: CREATION_DATE, direction: DESC }) {
    edges {
      node {
        checkoutId
      }
    }
  }

  checkouts(first: 20 filter: {
    /* NOT POSSIBLE TO EXCLUDE THE checkoutIds ABOVE? */ 
  }) {
    edges {
      node {
        email
        user {
          firstName
          lastName
        }
        id
        created
      }
    }
     pageInfo {
      hasNextPage
      hasPreviousPage
      startCursor
      endCursor
    }
    totalCount
  }
}
```

**Comment by peteror** [comment]
I see a couple of workarounds:

- On older Saleor versions checkout objects are deleted when they are converted to an order, so just querying checkouts may be a way. 
- Another way is I'm filtering checkout objects that are created before some threshold (depending on how quickly your orders are placed - this is what I do for the time being
- if nothing else works, you can just run a query on the Saleor database ¯\_(ツ)_/¯

**Comment by timuric** [comment]
@heyarne when checkout is converted to an order it is being deleted, therefore you can assume any checkout object is incomplete order
